### ISSUE TITLE
Responses eats all other Exceptions

### ISSUE THREAD

**Issue by Nashadelic** [getsentry/responses#72]
Suppose I setup responses for a url and then a call a function where the url is being requests. If there is an exception in that URL, responses always throws an exception with AssertionError: Not all requests has been executed

This becomes very hard to debug when this isn't really the exception. Is there a better way to set this up?

**Comment by ticosax** [comment]
One possible answer is #55

**Comment by javins** [comment]
Yikes! Just hit this one when one of my test cases errored out before it got to the intended end of a `RequestMock`'s context.  I spent the 30 minutes scratching my head before realizing that 'AssertionError: Not all requests have been executed' was a red herring and the error was in the test code.  I haven't yet given @ticosax's patch a try, but I'll give it a run tomorrow.

**Comment by pi-bjl** [comment]
Well there is an open PR from 2015 (#92), but here is my minimal reproduction in pytest. test_foo1 should show the ValueError, not the responses AssertionError. test_foo2 and test_foo3 have output as expected.

Test file: 

```
import responses

def foo(a):
    import requests
    if a == 1:
        raise ValueError()
    r = requests.get('http://localhost:8000/bar')
    if a == 2:
        raise IndexError()
    return r.text.upper()

def test_foo1():
    with responses.RequestsMock(assert_all_requests_are_fired=True) as rsps:
        rsps.add(responses.GET, 'http://localhost:8000/bar', body='baz')
        assert foo(1) == 'BAZ'

def test_foo2():
    with responses.RequestsMock(assert_all_requests_are_fired=True) as rsps:
        rsps.add(responses.GET, 'http://localhost:8000/bar', body='baz')
        assert foo(2) == 'BAZ'

def test_foo3():
    with responses.RequestsMock(assert_all_requests_are_fired=True) as rsps:
        rsps.add(responses.GET, 'http://localhost:8000/bar', body='baz')
        assert foo(3) == 'BAZ'
```

Test output:

```
$ python -m pytest -k test_foo -v --tb=short
================================== test session starts ===================================
platform linux2 -- Python 2.7.12, pytest-3.0.7, py-1.4.31, pluggy-0.4.0 -- /home/bjl/venv/bin/python
cachedir: .cache
rootdir: /home/bjl/workspace/, inifile:
plugins: xdist-1.16.0, flask-0.10.0, catchlog-1.2.2, celery-4.1.0
collected 158 items 

tests/test_foo.py::test_foo1 FAILED
tests/test_foo.py::test_foo2 FAILED
tests/test_foo.py::test_foo3 PASSED

======================================== FAILURES ========================================
_______________________________________ test_foo1 ________________________________________
tests/test_foo.py:21: in test_foo1
    assert foo(1) == 'BAZ'
../../../utils/venv/local/lib/python2.7/site-packages/responses.py:181: in __exit__
    self.stop()
../../../utils/venv/local/lib/python2.7/site-packages/responses.py:304: in stop
    [(url['method'], url['url']) for url in self._urls]))
E   AssertionError: Not all requests have been executed [(u'GET', 'http://localhost:8000/bar')]
_______________________________________ test_foo2 ________________________________________
tests/test_foo.py:27: in test_foo2
    assert foo(2) == 'BAZ'
tests/test_foo.py:13: in foo
    raise IndexError()
E   IndexError
================================== 155 tests deselected ==================================
=================== 2 failed, 1 passed, 155 deselected in 0.14 seconds ===================

```

**Comment by markstory** [comment]
I'm pretty sure this was fixed with 9d11c4d9
